src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -0,0 +1,678 @@
+                }
+            }
+            if (parsedMediaType.HasValue)
+            {
+                return new MediaTypeSegmentWithQuality(
+                    new StringSegment(mediaType, start, enumerator.CurrentOffset - start),
+                    quality);
+            }
+            else
+            {
+                return default(MediaTypeSegmentWithQuality);
+            }
+        }
+
+        private static Encoding GetEncodingFromCharset(StringSegment charset)
+        {
+            if (charset.Equals("utf-8", StringComparison.OrdinalIgnoreCase))
+            {
+                // This is an optimization for utf-8 that prevents the Substring caused by
+                // charset.Value
+                return Encoding.UTF8;
+            }
+
+            try
+            {
+                // charset.Value might be an invalid encoding name as in charset=invalid.
+                // For that reason, we catch the exception thrown by Encoding.GetEncoding
+                // and return null instead.
+                return charset.HasValue ? Encoding.GetEncoding(charset.Value) : null;
+            }
+            catch (Exception)
+            {
+                return null;
+            }
+        }
+
+        private static string CreateMediaTypeWithEncoding(StringSegment mediaType, Encoding encoding)
+        {
+            return $"{mediaType.Value}; charset={encoding.WebName}";
+        }
+
+        private bool ContainsAllParameters(Enumerator setParameters)
+        {
+            var parameterFound = true;
+            while (setParameters.MoveNext() && parameterFound)
+            {
+                if (setParameters.Current.HasName("q"))
+                {
+                    // "q" and later parameters are not involved in media type matching. Quoting the RFC: The first
+                    // "q" parameter (if any) separates the media-range parameter(s) from the accept-params.
+                    break;
+                }
+
+                // Copy the enumerator as we need to iterate multiple times over it.
+                // We can do this because it's a struct
+                var subsetParameters = _parameterEnumerator;
+                parameterFound = false;
+                while (subsetParameters.MoveNext() && !parameterFound)
+                {
+                    parameterFound = subsetParameters.Current.Equals(setParameters.Current);
+                }
+            }
+
+            return parameterFound;
+        }
+
+        private Enumerator GetEnumerator()
+        {
+            return new Enumerator(_mediaType, _offset, _length);
+        }
+
+        private struct Enumerator : IEnumerator<MediaTypeComponent>
+        {
+            private ParsingStatus _parsingStatus;
+            private string _mediaType;
+            private int _initialOffset;
+            private int _currentOffset;
+            private int? _length;
+            private MediaTypeComponent _current;
```

Comment:
This is not passed around through the stack, nor is MediaType, now we keep any of those two types in a list or something like that, is that your concern? I don't think there's an easy way to trim this down, at least none that I can come up with, do you have a suggestion that doesn't involve heavily reworking the current implementation?